File: temoto_nl_interface/actions/find_coordinates/src/getcoord_py/getcoord_nonTraversable_generation.py
import numpy as np
import cv2

def get(map_img, resolution, origin, robot_position):
    """
    Modifies the map image by setting all non-traversable areas (not reachable from the robot's position)
    to black (0, 0, 0), without altering the colors of traversable areas.

    Args:
        map_img (numpy.ndarray): The input map image (color image).
        resolution (float): The map's resolution in meters per pixel.
        origin (tuple): The (x, y) world coordinates of the map's origin.
        robot_position: The robot's position with world coordinates.

    Returns:
        numpy.ndarray: The modified map image with unreachable areas set to black.
    """
    # Extract robot's position in world coordinates
    robot_x = robot_position.transform.translation.x
    robot_y = robot_position.transform.translation.y

    # Map size in pixels
    height, width = map_img.shape[:2]

    # Convert robot's position to pixel coordinates
    origin_x, origin_y = origin[0], origin[1]
    pixel_x = int((robot_x - origin_x) / resolution)
    pixel_y = height - int((robot_y - origin_y) / resolution) - 1

    # Ensure pixel_x and pixel_y are within image bounds
    pixel_x = np.clip(pixel_x, 0, width - 1)
    pixel_y = np.clip(pixel_y, 0, height - 1)

    # Create free space mask: pixels with values 254 or 255 in all channels are free space
    free_space_mask = (
        np.all(map_img == [254, 254, 254], axis=2) |
        np.all(map_img == [255, 255, 255], axis=2)
    ).astype(np.uint8)

    # Prepare image for flood fill
    flood_fill_img = free_space_mask.copy()

    # Perform flood fill starting from the robot's pixel position
    # We fill the area with the value 2
    cv2.floodFill(flood_fill_img, None, (pixel_x, pixel_y), 2)

    # Create a mask of unreachable free spaces: areas not filled by flood fill
    unreachable_mask = (flood_fill_img != 2) & (free_space_mask == 1)

    # Set unreachable free space areas to black in the original map image
    map_img[unreachable_mask] = [0, 0, 0]

    return map_img

File: temoto_nl_interface/actions/find_coordinates/src/get_coordinates_scripts/getcoord_nonTraversable_generation.cpp
#include <opencv2/opencv.hpp>
#include <cmath>

cv::Mat get(const cv::Mat& map_img, double resolution, 
            const std::pair<double, double>& origin, 
            const struct RobotPosition& robot_position) {
    // Create a mutable copy of the map image
    cv::Mat modified_map = map_img.clone();

    // Extract robot's position in world coordinates
    double robot_x = robot_position.transform.translation.x;
    double robot_y = robot_position.transform.translation.y;

    // Map size in pixels
    int height = map_img.rows;
    int width = map_img.cols;

    // Convert robot's position to pixel coordinates
    double origin_x = origin.first;
    double origin_y = origin.second;
    
    int pixel_x = static_cast<int>((robot_x - origin_x) / resolution);
    int pixel_y = height - static_cast<int>((robot_y - origin_y) / resolution) - 1;

    // Ensure pixel_x and pixel_y are within image bounds
    pixel_x = std::max(0, std::min(pixel_x, width - 1));
    pixel_y = std::max(0, std::min(pixel_y, height - 1));

    // Create free space mask
    cv::Mat free_space_mask = cv::Mat::zeros(height, width, CV_8UC1);

    // Identify free space pixels (254 or 255 in all channels)
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            cv::Vec3b pixel = map_img.at<cv::Vec3b>(y, x);
            if ((pixel[0] == 254 && pixel[1] == 254 && pixel[2] == 254) ||
                (pixel[0] == 255 && pixel[1] == 255 && pixel[2] == 255)) {
                free_space_mask.at<uchar>(y, x) = 1;
            }
        }
    }

    // Prepare image for flood fill
    cv::Mat flood_fill_img = free_space_mask.clone();

    // Perform flood fill starting from the robot's pixel position
    cv::floodFill(flood_fill_img, cv::Mat(), cv::Point(pixel_x, pixel_y), 2);

    // Create a mask of unreachable free spaces
    cv::Mat unreachable_mask = cv::Mat::zeros(height, width, CV_8UC1);
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            if (free_space_mask.at<uchar>(y, x) == 1 && 
                flood_fill_img.at<uchar>(y, x) != 2) {
                unreachable_mask.at<uchar>(y, x) = 1;
            }
        }
    }

    // Set unreachable free space areas to black
    for (int y = 0; y < height; ++y) {
        for (int x = 0; x < width; ++x) {
            if (unreachable_mask.at<uchar>(y, x) == 1) {
                modified_map.at<cv::Vec3b>(y, x) = cv::Vec3b(0, 0, 0);
            }
        }
    }

    return modified_map;
}

// Struct to represent robot position (you may need to adjust based on your actual implementation)
struct RobotPosition {
    struct {
        struct {
            double x;
            double y;
        } translation;
    } transform;
};